### PR TITLE
[sdl3] Adding port for SDL3 Pre-release

### PR DESCRIPTION
**Pull request by Honeybunch**
Citing https://github.com/microsoft/vcpkg/issues/32062 and the fact that the new SDL3 MkII GPU API [got some buzz](https://news.ycombinator.com/item?id=41396260) I thought people might find an SDL3 port useful. I've been using this port myself for a little while so figured I'd try up-streaming it. Should be easy to update this port for the full SDL 3.0.0 release

Note: SDL3 has dropped support for UWP so it has been removed from `supports` 😞

- [X] Changes comply with the [maintainer guide](https://github.com/microsoft/vcpkg-docs/blob/main/vcpkg/contributing/maintainer-guide.md).
- [X] The name of the port matches an existing name for this component on https://repology.org/ if possible, and/or is strongly associated with that component on search engines.
- [X] Optional dependencies are resolved in exactly one way. For example, if the component is built with CMake, all `find_package` calls are REQUIRED, are satisfied by `vcpkg.json`'s declared dependencies, or disabled with [CMAKE_DISABLE_FIND_PACKAGE_Xxx](https://cmake.org/cmake/help/latest/variable/CMAKE_DISABLE_FIND_PACKAGE_PackageName.html).
- [X] The versioning scheme in `vcpkg.json` matches what upstream says.
- [X] The license declaration in `vcpkg.json` matches what upstream says.
- [X] The installed as the "copyright" file matches what upstream says.
- [X] The source code of the component installed comes from an authoritative source.
- [X] The generated "usage text" is accurate. See [adding-usage](https://github.com/microsoft/vcpkg-docs/blob/main/vcpkg/examples/adding-usage.md) for context.
- [X] The version database is fixed by rerunning `./vcpkg x-add-version --all` and committing the result.
- [X] Only one version is in the new port's versions file.
- [X] Only one version is added to each modified port's versions file.

### CODE REVIEW

**Comment by LilyWangLL** [comment]
All features passed with `x64-linux`, usage test passed on `x64-windows`.

**Review by LilyWangLL — DISMISSED** [review]

**Comment by BillyONeal** [comment]
I confirmed that SDL2 and SDL3 can be concurrently installed.

**Review by BillyONeal — DISMISSED** [review]
I think the port is good, thanks for the new port!

I would like to double check with other maintainers that we are OK with merging a port with no upstream releases yet; they might not be wanting people to treat it as generally available, and publishing it in vcpkg might do that. To that end it might be good to use version-string and the git SHA instead, as ensuring reasonable sort order with other prerelease commits sounds questionable.

If there's a statement from upstream maintainers that they're happy with this that avoids all concerns.

Thanks again!

**Comment by Honeybunch** [comment]
Totally! Ill file an issue on their repo and see what they say 😁

The git SHA is a good idea I can do that.

**Comment by BillyONeal** [comment]
In https://github.com/libsdl-org/SDL/issues/10792 , upstream maintainer @icculus says:

>At this point, I'd say please wait until the ABI lockdown, which is coming soon.
>After that, we might add stuff but we will not change or remove existing stuff, so people will be able to update a package without their code breaking.

**Comment by duerdak** [comment]
This can be merged now that the ABI lock has happened.

> We are live!
> 
> https://github.com/libsdl-org/SDL/releases/tag/preview-3.1.3
> 
> Feel free to package it!

**Comment by Honeybunch** [comment]
Yep. My goal is to get this port updated and ready for review today or tomorrow.

**Comment by Honeybunch** [comment]
My downstream project uses this successfully

https://github.com/HoneybunchBuilder/toybox/pull/12

File: ports/sdl3/portfile.cmake [hunk 1]
```diff
@@ -0,0 +1,88 @@
...
+    message(WARNING "You will need to install Wayland dependencies to use feature wayland:\nsudo apt install libwayland-dev libxkbcommon-dev libegl1-mesa-dev\n")
+endif()
+if ("ibus" IN_LIST FEATURES)
+    message(WARNING "You will need to install ibus dependencies to use feature ibus:\nsudo apt install libibus-1.0-dev\n")
+endif()
+
+vcpkg_cmake_configure(
+    SOURCE_PATH "${SOURCE_PATH}"
+    ${configure_opts}
+    OPTIONS ${FEATURE_OPTIONS}
```
Comment (dg0yt) [inline ports/sdl3/portfile.cmake:35]:
```suggestion
    OPTIONS
        ${FEATURE_OPTIONS}
```

File: ports/sdl3/portfile.cmake [hunk 2]
```diff
@@ -0,0 +1,88 @@
...
+        -DSDL_STATIC=${SDL_STATIC}
+        -DSDL_SHARED=${SDL_SHARED}
+        -DSDL_FORCE_STATIC_VCRT=${FORCE_STATIC_VCRT}
+        -DSDL_LIBC=ON
+        -DSDL_TEST_LIBRARY=OFF
+        -DSDL_TESTS=OFF
+        -DSDL_INSTALL_CMAKEDIR="cmake"
```
Comment (dg0yt) [inline ports/sdl3/portfile.cmake:42]: No inner quotes here.
Does this controls cmake config destination? (I don't know.)

File: ports/sdl3/portfile.cmake [hunk 3]
```diff
@@ -0,0 +1,88 @@
...
+        SDL_FORCE_STATIC_VCRT
+        PKG_CONFIG_USE_CMAKE_PREFIX_PATH
+)
+
+vcpkg_cmake_install()
+if(VCPKG_TARGET_IS_WINDOWS)
+    vcpkg_cmake_config_fixup(CONFIG_PATH cmake)
+else()
+    vcpkg_cmake_config_fixup(CONFIG_PATH lib/cmake/SDL3)
+endif()
```
Comment (dg0yt) [inline ports/sdl3/portfile.cmake:55]: So SDL_INSTALL_CMAKEDIR didn't change it?

File: ports/sdl3/portfile.cmake [hunk 4]
```diff
@@ -0,0 +1,88 @@
...
+
+vcpkg_copy_pdbs()
+
+set(DYLIB_COMPATIBILITY_VERSION_REGEX "set\\(DYLIB_COMPATIBILITY_VERSION (.+)\\)")
+set(DYLIB_CURRENT_VERSION_REGEX "set\\(DYLIB_CURRENT_VERSION (.+)\\)")
+file(STRINGS "${SOURCE_PATH}/CMakeLists.txt" DYLIB_COMPATIBILITY_VERSION REGEX ${DYLIB_COMPATIBILITY_VERSION_REGEX})
+file(STRINGS "${SOURCE_PATH}/CMakeLists.txt" DYLIB_CURRENT_VERSION REGEX ${DYLIB_CURRENT_VERSION_REGEX})
+string(REGEX REPLACE ${DYLIB_COMPATIBILITY_VERSION_REGEX} "\\1" DYLIB_COMPATIBILITY_VERSION "${DYLIB_COMPATIBILITY_VERSION}")
+string(REGEX REPLACE ${DYLIB_CURRENT_VERSION_REGEX} "\\1" DYLIB_CURRENT_VERSION "${DYLIB_CURRENT_VERSION}")
```
Comment (dg0yt) [inline ports/sdl3/portfile.cmake:83]: This seems to have no effect.

Comment (Honeybunch) [reply ports/sdl3/portfile.cmake:83]: Oh yeah I didn't do this part originally. It can probably be nuked

File: ports/sdl3/portfile.cmake (hunk 2 above)
Comment (Honeybunch) [reply ports/sdl3/portfile.cmake:42]: That was the idea but taking another look the sdl cmake script continues to mutate this so it's a wash

File: ports/sdl3/portfile.cmake (hunk 3 above)
Comment (Honeybunch) [reply ports/sdl3/portfile.cmake:55]: Nope that's my bad. Good catch, thank you

File: ports/sdl3/usage [hunk 5]
```diff
@@ -0,0 +1,7 @@
+sdl3 provides CMake targets:
+
+    find_package(SDL3 CONFIG REQUIRED)
+    target_link_libraries(main
+        PRIVATE
+        $<IF:$<TARGET_EXISTS:SDL3::SDL3>,SDL3::SDL3,SDL3::SDL3-static>
```
Comment (madebr) [inline ports/sdl3/usage:6]: https://github.com/libsdl-org/SDL/blob/main/docs/README-cmake.md#a-system-sdl-library

```suggestion
        $<IF:$<TARGET_EXISTS:SDL3::SDL3-shared>,SDL3::SDL3-shared,SDL3::SDL3-static>
```

File: ports/sdl3/portfile.cmake [hunk 6]
```diff
@@ -0,0 +1,79 @@
...
+    vcpkg_cmake_config_fixup(CONFIG_PATH cmake)
+else()
+    vcpkg_cmake_config_fixup(CONFIG_PATH lib/cmake/SDL3)
+endif()
+
+file(REMOVE_RECURSE
+    "${CURRENT_PACKAGES_DIR}/debug/include"
+    "${CURRENT_PACKAGES_DIR}/debug/share"
+    "${CURRENT_PACKAGES_DIR}/bin/sdl3-config"
+    "${CURRENT_PACKAGES_DIR}/debug/bin/sdl3-config"
```
[the remark this answers is not in the document]
Comment (Honeybunch) [reply ports/sdl3/portfile.cmake:60]: Oh thank you good catch

File: versions/s-/sdl3.json [hunk 7]
```diff
@@ -0,0 +1,14 @@
...
+    {
+      "git-tree": "1e59c397c1417a9073c8eeb6e2672c48fdf5e2cb",
+      "version": "3.1.3-preview",
+      "port-version": 0
+    },
+    {
+      "git-tree": "9a9900fcfe66dbbbc9b1dacd66e54ae9a1d90a18",
+      "version": "3.0.0-pre",
+      "port-version": 0
+    }
```
[the remark this answers is not in the document]
Comment (Honeybunch) [reply versions/s-/sdl3.json:12]: Ooo yep that's an easy one

File: ports/sdl3/usage (hunk 5 above)
Comment (Honeybunch) [reply ports/sdl3/usage:6]: :+1:

File: ports/sdl3/portfile.cmake [hunk 8]
```diff
@@ -0,0 +1,77 @@
...
+    "${CURRENT_PACKAGES_DIR}/share/aclocal"
+)
+
+file(GLOB BINS "${CURRENT_PACKAGES_DIR}/debug/bin/*" "${CURRENT_PACKAGES_DIR}/bin/*")
+if(NOT BINS)
+    file(REMOVE_RECURSE
+        "${CURRENT_PACKAGES_DIR}/bin"
+        "${CURRENT_PACKAGES_DIR}/debug/bin"
+    )
+endif()
```
Comment (dg0yt) [inline ports/sdl3/portfile.cmake:71]: That't a point where I always ask: What is removed here? 
If there are tools, couldn't it be handle with `vcpkg_copy_tools(... AUTO_CLEAN)`?

If there are tools and DLLs, the tools will not be removed.
If there are no tools and no DLLs, the directories shouldn't be created.

File: ports/sdl3/portfile.cmake [hunk 9]
```diff
@@ -0,0 +1,77 @@
...
+else()
+    vcpkg_cmake_config_fixup(CONFIG_PATH lib/cmake/SDL3)
+endif()
+
+file(REMOVE_RECURSE
+    "${CURRENT_PACKAGES_DIR}/debug/include"
+    "${CURRENT_PACKAGES_DIR}/debug/share"
+    "${CURRENT_PACKAGES_DIR}/SDL3.framework"
+    "${CURRENT_PACKAGES_DIR}/debug/SDL3.framework"
+    "${CURRENT_PACKAGES_DIR}/share/licenses"
```
Comment (dg0yt) [inline ports/sdl3/portfile.cmake:61]: I assume you checked that it is sufficiently covered in the installed `copyright`.

File: ports/sdl3/portfile.cmake [hunk 10]
```diff
@@ -0,0 +1,77 @@
...
+    "${CURRENT_PACKAGES_DIR}/share/aclocal"
```
Comment (dg0yt) [inline ports/sdl3/portfile.cmake:62]: Is there a problem with these files?

Comment (madebr) [reply ports/sdl3/portfile.cmake:62]: SDL3 does not ship autoconf macros anymore.
```suggestion
```

Comment (Honeybunch) [reply ports/sdl3/portfile.cmake:62]: Legacy from the SDL2 port

File: ports/sdl3/portfile.cmake (hunk 8 above)
Comment (Honeybunch) [reply ports/sdl3/portfile.cmake:71]: Legacy from the SDL2 port. Not sure what's up with it. Let's remove it and see!

File: ports/sdl3/usage [hunk 11]
```diff
@@ -0,0 +1,7 @@
+sdl3 provides CMake targets:
+
+    find_package(SDL3 CONFIG REQUIRED)
+    target_link_libraries(main
+        PRIVATE
+        $<IF:$<TARGET_EXISTS:SDL3::SDL3-shared>,SDL3::SDL3-shared,SDL3::SDL3-static>
```
Comment (NekkoDroid) [inline ports/sdl3/usage:6]: Is there a particular reason that this is done over just linking directly to `SDL3::SDL3`? The `SDL3Config.cmake` already resolves for the correct version and aliases that to `SDL3::SDL3`.

**Comment by Honeybunch** [comment]
I don't think this needs to be tagged as depending on upstream changes anymore

**Review by LilyWangLL — DISMISSED** [review]

File: ports/sdl3/vcpkg.json [hunk 12]
```diff
@@ -0,0 +1,63 @@
...
+      "platform": "linux"
+    },
+    {
+      "name": "wayland",
+      "platform": "linux"
+    },
+    {
+      "name": "x11",
+      "platform": "linux"
+    }
```
Comment (JavierMatosD) [inline ports/sdl3/vcpkg.json:35]: @Honeybunch, should these be default-features? Please see, https://learn.microsoft.com/en-us/vcpkg/contributing/maintainer-guide#default-features-should-enable-behaviors-not-apis

Meaning, that it is very rare for something to belong to default-features. These all seem to enable API's.

Comment (JavierMatosD) [inline ports/sdl3/vcpkg.json:35]:
```suggestion

```

Comment (Honeybunch) [reply ports/sdl3/vcpkg.json:35]: These are the same defaults that SDL2 used

Comment (madebr) [reply ports/sdl3/vcpkg.json:35]: If you disable these by default, the default sdl2 package will now display x11/wayland windows on Linux.

Comment (Honeybunch) [reply ports/sdl3/vcpkg.json:35]: Also, yes. On Linux you want ibus, wayland and x11 by default. At least IMO

**Comment by Honeybunch** [comment]
Does this require any other changes?

**Review by BillyONeal — DISMISSED** [review]
@JavierMatosD Do you know why you asked for team review?

**Comment by JavierMatosD** [comment]
> @JavierMatosD Do you know why you asked for team review?

I'm asking for team review since these default features seem to enable APIs which is against our current policy. We accepted the sdl2 port's feature but I _think_ this was done erroneously or the policy was different when it was added.

**Comment by maia-s** [comment]
As of <https://github.com/libsdl-org/SDL/pull/11310>, SDL fails to build on Linux if neither X11 or Wayland is enabled

**Comment by Honeybunch** [comment]
Any ETA on this?

**Comment by zby-c** [comment]
hurry up 😍 i need this

File: ports/sdl3/portfile.cmake [hunk 13]
```diff
@@ -0,0 +1,64 @@
+vcpkg_from_github(
+    OUT_SOURCE_PATH SOURCE_PATH
+    REPO libsdl-org/SDL
+    REF "preview-3.1.3"
+    SHA512 60c489fbec3e9f5cbde57520edf20257c27facab3c44e8e9b719743ee5035e80221eb9403016bfbc696dfab834f955a81569539b3859171afafb8970b3a7a4dd
```
Comment (madebr) [inline ports/sdl3/portfile.cmake:5]:
```suggestion
    REF "preview-3.1.6"
    SHA512 80d3a0cee9a17d78becaee5bbc876408dabc865ec2ec18dec5d8c822c0ca523faab5384164e919d3b66ddd487f05a86b562bbb98bc2f677dd2b9081439a5fa5e
```

File: versions/s-/sdl3.json [hunk 14]
```diff
@@ -0,0 +1,9 @@
+{
+  "versions": [
+    {
+      "git-tree": "71afeeba55c0d1eb02027f8a867757b0e6924d3d",
+      "version": "3.1.3-preview",
```
Comment (madebr) [inline versions/s-/sdl3.json:5]:
```suggestion
      "git-tree": "78cc5c173404488d80751af226d1eaf67033bcc4",
      "version": "3.1.6-preview",
```

**Review by madebr — COMMENTED** [review]
https://github.com/libsdl-org/SDL/releases/tag/preview-3.1.6

**Comment by zby-c** [comment]
MERGE plz 🔥☺️

File: ports/sdl3/portfile.cmake (hunk 13 above)
Comment (LilyWangLL) [reply ports/sdl3/portfile.cmake:5]: Ping @Honeybunch for response.

**Review by LilyWangLL — APPROVED** [review]

**Review by vicroms — APPROVED** [review]

**Review by zby-c — APPROVED** [review]

**Comment by Pentarctagon** [comment]
To go along with this - are there any plans to add the SDL3 version of SDL_mixer and SDL_image?

**Comment by Honeybunch** [comment]
I do have an sdl3 mixer port I can try to upstream. Not sure if that or SDL3_Image have had an ABI lock and a prerelease tag cut though

**Comment by Honeybunch** [comment]
I do not have an SDL3_image port since i havent needed it

**Comment by madebr** [comment]
Only SDL3 is in API/ABI lock. None of SDL's satellite libraries are considered stable at this moment.

**Comment by Honeybunch** [comment]
Thanks for the clarification 😁

**Comment by dg0yt** [comment]
This new port has uncontrolled dependencies with installation order problems. For example, it fails with hidapi and static libusb installed on android. This affects vcpkg CI. 

There may be more issues, but due to `pkg_check_modules` it is difficult to track and control. 

I also see copies of khronos headers which have ports in vcpkg.

**Comment by pbdot** [comment]
The import prefix fails on ubuntu, because the config cmake is in `vcpkg\packages\sdl3_x64-linux\share\sdl3\SDL3` vs on Windows `vcpkg\packages\sdl3_x64-windows-static\share\sdl3` meaning that it is looking for the required include and lib files in `share` and not the actual package folder. 

```# Compute the installation prefix relative to this file.
get_filename_component(_IMPORT_PREFIX "${CMAKE_CURRENT_LIST_FILE}" PATH)
get_filename_component(_IMPORT_PREFIX "${_IMPORT_PREFIX}" PATH)
get_filename_component(_IMPORT_PREFIX "${_IMPORT_PREFIX}" PATH)```

**Comment by dg0yt** [comment]
> The import prefix fails on ubuntu, because the config cmake is in `vcpkg\packages\sdl3_x64-linux\share\sdl3\SDL3`

That's the wrong location, indeed. But my diagnose is a little bit different. https://github.com/microsoft/vcpkg/pull/42206.

**Pull request merged**